File: Tag/Dynamic-Programming/Easy/868.二进制间距.java
/*
 * @lc app=leetcode.cn id=868 lang=java
 *
 * [868] 二进制间距
 */

// @lc code=start
class Solution {
    public int binaryGap(int n) { // 时间复杂度：O(logn)
        int max = 0;
        int count = 0;
        while ((n & 1) == 0) {
            n >>= 1;
        }
        
        do {
            count ++;
            n = n >> 1;
            if ((n & 1) == 1) {
                max = Math.max(max, count);
                count = 0;
            }
        } while(n > 0);

        return max;
    }
}
// @lc code=end

File: jsTest/dynamic-programming/easy/868.二进制间距.js
/*
 * @lc app=leetcode.cn id=868 lang=javascript
 *
 * [868] 二进制间距
 */

// @lc code=start
/**
 * @param {number} n
 * @return {number}
 */
var binaryGap = function(n) { // 时间复杂度 O(logn)
    let maxGap = 0;
    let gap = 0;
    while((n & 1) == 0) {
        n = n >> 1
    }

    do{
        gap += 1
        n  = n >> 1
        if ((n & 1) == 1) {
            maxGap = Math.max(maxGap, gap)
            gap = 0
        }
    } while (n > 0)
    
    return maxGap
};
// @lc code=end

